File: server/kafka-backend/src/infrastructure/services/retry.service.ts
import { Injectable, Logger } from '@nestjs/common';

export enum BackoffStrategy
{
  LINEAR = 'linear',
  EXPONENTIAL = 'exponential',
  DECORRELATED_JITTER = 'decorrelated_jitter',
}

export interface RetryOptions
{
  maxAttempts: number;
  baseDelayMs: number;
  maxDelayMs: number;
  strategy: BackoffStrategy;
  retryableErrors?: Array<string | RegExp>;
}

const DEFAULT_OPTIONS: RetryOptions = {
  maxAttempts: 3,
  baseDelayMs: 1000,
  maxDelayMs: 30000,
  strategy: BackoffStrategy.EXPONENTIAL,
};

@Injectable()
export class RetryService
{
  private readonly logger = new Logger(RetryService.name);

  async execute<T>(
    operation: () => Promise<T>,
    options?: Partial<RetryOptions>,
  ): Promise<T>
  {
    const opts = { ...DEFAULT_OPTIONS, ...options };
    let lastError: Error | undefined;

    for(let attempt = 1; attempt <= opts.maxAttempts; attempt++)
    {
      try
      {
        return await operation();
      }
      catch(err)
      {
        lastError = err instanceof Error ? err : new Error(String(err));

        // Propagate immediately for non-retryable failures instead of wasting attempts
        if(!this.isRetryable(lastError, opts.retryableErrors))
        {
          throw lastError;
        }

        if(attempt >= opts.maxAttempts)
        {
          throw lastError;
        }

        const delay = this.calculateDelay(attempt, opts);

        this.logger.warn(
          `Attempt ${attempt}/${opts.maxAttempts} failed: ${lastError.message}. ` +
          `Retrying in ${delay}ms...`,
        );

        await this.sleep(delay);
      }
    }

    throw lastError ?? new Error('Retry exhausted with unknown error');
  }

  async executeSafe<T>(
    operation: () => Promise<T>,
    options?: Partial<RetryOptions>,
  ): Promise<{ success: true; value: T } | { success: false; error: Error }>
  {
    try
    {
      const value = await this.execute(operation, options);
      return { success: true, value };
    }
    catch(err)
    {
      return {
        success: false,
        error: err instanceof Error ? err : new Error(String(err)),
      };
    }
  }

  private calculateDelay(attempt: number, options: RetryOptions): number
  {
    switch(options.strategy)
    {
      case BackoffStrategy.LINEAR:
        return Math.min(options.baseDelayMs * attempt, options.maxDelayMs);

      case BackoffStrategy.EXPONENTIAL:
        return Math.min(
          options.baseDelayMs * Math.pow(2, attempt - 1),
          options.maxDelayMs,
        );

      case BackoffStrategy.DECORRELATED_JITTER:
      {
        // Prevents thundering herd when multiple consumers retry simultaneously
        // sleep = min(cap, random_between(base, sleep * 3))
        const cap = options.maxDelayMs;
        const base = options.baseDelayMs;
        const prevDelay = options.baseDelayMs * Math.pow(2, attempt - 2);
        const range = Math.min(cap, Math.max(base, prevDelay * 3));
        return base + Math.random() * (range - base);
      }

      default:
        return options.baseDelayMs;
    }
  }

  // Defaults to retryable for all errors unless explicit filter provided
  private isRetryable(
    error: Error,
    retryableErrors?: Array<string | RegExp>,
  ): boolean
  {
    if(!retryableErrors || retryableErrors.length === 0)
    {
      return true;
    }

    const message = error.message;
    return retryableErrors.some((pattern) =>
      typeof pattern === 'string'
        ? message.includes(pattern)
        : pattern.test(message),
    );
  }

  private sleep(ms: number): Promise<void>
  {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: server/java-backend/src/main/java/com/intbank/service/RetryService.java
package com.intbank.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class RetryService
{

    private static final Logger log = LoggerFactory.getLogger(RetryService.class);

    private static final RetryOptions DEFAULT_OPTIONS = new RetryOptions(3, 1000, 30000, BackoffStrategy.EXPONENTIAL, List.of());

    public <T> T execute(RetryableOperation<T> operation, RetryOptions options) throws Exception
    {
        RetryOptions opts = options != null ? options : DEFAULT_OPTIONS;
        Exception lastError = null;

        for (int attempt = 1; attempt <= opts.maxAttempts(); attempt++) {
            try {
                return operation.execute();
            } catch (Exception err) {
                lastError = err;
                if (!isRetryable(err, opts.retryableErrors())) {
                    throw err;
                }
                if (attempt >= opts.maxAttempts()) {
                    throw err;
                }
                long delay = calculateDelay(attempt, opts);
                log.warn("Attempt {}/{} failed: {}. Retrying in {}ms...", attempt, opts.maxAttempts(), err.getMessage(), delay);
                Thread.sleep(delay);
            }
        }
        throw lastError != null ? lastError : new RuntimeException("Retry exhausted with unknown error");
    }

    private boolean isRetryable(Exception error, List<String> retryableErrors)
    {
        if (retryableErrors == null || retryableErrors.isEmpty()) return true;
        String msg = error.getMessage() != null ? error.getMessage() : "";
        return retryableErrors.stream().anyMatch(msg::contains);
    }

    private long calculateDelay(int attempt, RetryOptions opts)
    {
        return switch (opts.strategy()) {
            case LINEAR -> Math.min(opts.baseDelayMs() * attempt, opts.maxDelayMs());
            case EXPONENTIAL -> Math.min(opts.baseDelayMs() * (long) Math.pow(2, attempt - 1), opts.maxDelayMs());
            case DECORRELATED_JITTER -> {
                long prevDelay = opts.baseDelayMs() * (long) Math.pow(2, attempt - 2);
                long range = Math.min(opts.maxDelayMs(), Math.max(opts.baseDelayMs(), prevDelay * 3));
                yield opts.baseDelayMs() + (long) (Math.random() * (range - opts.baseDelayMs()));
            }
            default -> opts.baseDelayMs();
        };
    }

    public interface RetryableOperation<T> {
        T execute() throws Exception;
    }

    public record RetryOptions(int maxAttempts, long baseDelayMs, long maxDelayMs, BackoffStrategy strategy, List<String> retryableErrors) {}

    public enum BackoffStrategy { LINEAR, EXPONENTIAL, DECORRELATED_JITTER }
}